### ISSUE TITLE
Chatbot loss function with ignore_index instead of target's padding mask cause that model doesn't converge

### ISSUE THREAD

**Issue by radoslawkrolikowski** [pytorch/tutorials#663]
Hello.
@zhangguanheng66 
@SethHWeidman 

I was going through Pytorch official tutorial concerning the implementation of seq2eq Chatbot with Attention.  Actual solution that uses binary mask tensor describing the padding of the target tensor seems to be a little too complicated, thus I wanted to take advantage of loss function parameter `ingore_index = PAD_token` , that coerce mentioned function to omit specific target value, so that it doesn't contribute when calculating the loss.

> criterion = nn.NLLLoss(ignore_index=PAD_token)

In the definition of _train_ function I only made a small change in the loop when we forward propagate through decoder one step at a time:

```
    for t in range(max_target_len):
        decoder_output, decoder_hidden = decoder(
            decoder_input, decoder_hidden, encoder_outputs
        )
        # Teacher forcing: next input is current target
        decoder_input = target_variable[t].view(1, -1)

        # Calculate and accumulate loss
        # decoder_output: (batch_size, vocab_size)
        # target_variable[t]: (batch_size)
        loss = criterion(decoder_output, target_variable[t])
        total_loss += loss
        print_losses.append(loss.item())


# Perform backpropatation
total_loss.backward()
```

Now model doesn't converge at all. Total loss fluctuate near the same value, but it doesn’t want to decrease. I tried the following:

- used reduction='mean' while initializing loss function, so that the loss is averaged over non-ignored targets.
criterion = nn.NLLLoss(ignore_index=PAD_token, reduction='mean')

- used reduction='sum' together with:
```
            # Calculate the loss function
            loss = criterion(decoder_output, target_seq[t])
            
            # Count how many words there are in each batch (non-padding tokens)
            n_words = sum([1 if x>0 else 0 for x in target_seq[t]])
            total_loss += loss
            total_words += n_words
            losses.append(loss.item())
    # Perform backpropatation
    total_loss /= total_words
    total_loss.backward()
```

- used several learning rates from the range 1 - 0.00001

- used different optimizers

- checked if encoder.embedding.weight update properly

- used nn.CrossEntropyLoss

It seems like using ignore_index instead of padding mask makes the model untrainable.

Could anyone please explain how to use properly any kind of loss function with ignore_index in seq2seq models with variate length sequences? 

Best regards.

**Comment by radoslawkrolikowski** [comment]
I would like to present a simple comparison of the change of the loss during training between original tutorial code and code with only the change of the loss calculation method.

Calculation of loss using target's padding mask
```
for t in range(max_target_len-1):
        decoder_output, decoder_hidden = decoder(
            decoder_input, decoder_hidden, encoder_outputs)
        # Teacher forcing: next input is current target
        decoder_input = target_variable[t].view(1, -1)
        # Calculate and accumulate loss
        mask_loss, nTotal = maskNLLLoss(decoder_output, target_variable[t], mask[t])
        loss += mask_loss
        print_losses.append(mask_loss.item() * nTotal)
        n_totals += nTotal
loss.backward()

# Clip gradients: gradients are modified in place
_ = nn.utils.clip_grad_norm_(encoder.parameters(), clip)
_ = nn.utils.clip_grad_norm_(decoder.parameters(), clip)

# Adjust model weights
encoder_optimizer.step()
decoder_optimizer.step()
return sum(print_losses)/ n_totals
```
In the preceding code we are going to change only two lines of code: use `CrossEntropyLoss(ignore_index=PAD_token, reduction='mean')` instead of `maskNLLLoss(inp, target, mask)`
```
# USE NEW LOSS FUNCTION
mask_loss = criterion(decoder_output, target_variable[t])
loss += mask_loss
...
# AT THE END DIVIDE BY (max_target_len-1) 
return sum(print_losses)/(max_target_len-1) 
```

The training loss for model that uses padding mask:
Iteration: 1; Percent complete: 0.0%; Average loss: 8.9560
Iteration: 2; Percent complete: 0.1%; Average loss: 8.8288
Iteration: 3; Percent complete: 0.1%; Average loss: 8.6317
Iteration: 4; Percent complete: 0.1%; Average loss: 8.2510
Iteration: 5; Percent complete: 0.1%; Average loss: 7.9949
Iteration: 6; Percent complete: 0.1%; Average loss: 7.3578
Iteration: 7; Percent complete: 0.2%; Average loss: 6.9733
Iteration: 8; Percent complete: 0.2%; Average loss: 6.4963
Iteration: 9; Percent complete: 0.2%; Average loss: 6.5024
Iteration: 10; Percent complete: 0.2%; Average loss: 6.2277

The training loss for model that uses `CrossEntropyLoss(ignore_index=PAD_token, reduction='mean)`
Iteration: 1; Percent complete: 0.0%; Average loss: 8.9652
Iteration: 2; Percent complete: 0.1%; Average loss: 8.9649
Iteration: 3; Percent complete: 0.1%; Average loss: 8.8114
Iteration: 4; Percent complete: 0.1%; Average loss: 8.8202
Iteration: 5; Percent complete: 0.1%; Average loss: 8.8211
Iteration: 6; Percent complete: 0.1%; Average loss: 8.8219
Iteration: 7; Percent complete: 0.2%; Average loss: 8.8096
Iteration: 8; Percent complete: 0.2%; Average loss: 8.8063
Iteration: 9; Percent complete: 0.2%; Average loss: 8.8104
Iteration: 10; Percent complete: 0.2%; Average loss: 8.8138
Iteration: 11; Percent complete: 0.3%; Average loss: 8.7909
Iteration: 12; Percent complete: 0.3%; Average loss: 8.8143
Iteration: 13; Percent complete: 0.3%; Average loss: 8.8106
Iteration: 14; Percent complete: 0.4%; Average loss: 8.8270
Iteration: 15; Percent complete: 0.4%; Average loss: 8.8286
Iteration: 16; Percent complete: 0.4%; Average loss: 8.8267
Iteration: 17; Percent complete: 0.4%; Average loss: 8.8087
Iteration: 18; Percent complete: 0.4%; Average loss: 8.8398
Iteration: 19; Percent complete: 0.5%; Average loss: 8.8203
Iteration: 20; Percent complete: 0.5%; Average loss: 8.8226

We can see that for both methods the initial values of loss are almost equal, so I think it is sufficient evidence that loss using `criterion` is calculated properly. I was wondering if is that possible that backward() calculates completely different gradients based on the same loss given that this loss was calculated using various methods. For instance gradients `dloss1/dx` and dloss2/dx` would be of a different magnitude only because loss1 = loss2 were calculated using various functions? Going along chain (chain rule) we would have different partial derivatives but small change of for example embedding weights should affect the change of loss the same no matter which method we use, right?

I checked out the sum of the gradients of `encoder.embedding.weight.grad `during training for both models.
For model with padding mask the lower the loss was, the lower the gradients, but they decreased very slowly.
Iteration: 1; tensor(-0.1997)
Iteration: 35; tensor(-0.0057)

In contrast to preceding example, model with `criterion` begins with much lower gradients, and during the training gradients diminish much faster.
Iteration: 1; tensor(-5.2681e-05)
Iteration: 35; tensor(2.0964e-10)

Could anybody please explain this behavior? Why the same optimizer calculates so much different gradients? Any other advice would be appreciated as well. 

Best regards.
Radek.

**Comment by radoslawkrolikowski** [comment]
In order to deal with vanishing gradients I used batch normalization layer before `Softmax`, then I used tensorboardX to visualize weights and gradients during training. 
On the left example of encoder.gru weights and on the right encoder.gru gradients:
![image](https://user-images.githubusercontent.com/55454254/65427281-3f0b6600-de12-11e9-8283-b39ae5aaaa2a.png)
As we can see, still there is a problem with gradients (now they most commonly equal zero). 

I decided to one more time check everything out, including loss function, which as I thought returned reasonable values, but despite that it turned out that loss was not calculated properly, because I wrongly combined the use of loss function with softmax layer. Namely, using `CrossEntropyLoss` I should got rid of Softmax at all, as `CrossEntropyLoss` comprised of `LogSoftmax` and `NLLLoss`, and using `NLLLoss` I should changed `Softmax` to `LogSoftmax`.
After these changes loss decrease and model finally learns something.
